flang/lib/Lower/Bridge.cpp

```diff
@@ -517,14 +522,19 @@ class FirConverter : public Fortran::lower::AbstractConverter {
         [&sb](auto &) { return sb.toExtendedValue(); });
   }
 
-  mlir::Value createTemp(mlir::Location loc,
-                         const Fortran::semantics::Symbol &sym,
-                         llvm::ArrayRef<mlir::Value> shape = {}) {
-    // FIXME: should return fir::ExtendedValue
-    if (Fortran::lower::SymbolBox v = lookupSymbol(sym))
-      return v.getAddr();
-    mlir::Value newVal = builder->createTemporary(
-        loc, genType(sym), toStringRef(sym.name()), shape);
+  /// Generate the address of loop variable \p sym.
+  /// If \p sym is not mapped yet, allocate local storage for it.
+  mlir::Value genLoopVariableAddress(mlir::Location loc,
+                                     const Fortran::semantics::Symbol &sym) {
+    if (lookupSymbol(sym)) {
+      Fortran::lower::StatementContext stmtCtx;
+      return fir::getBase(
+          genExprAddr(Fortran::evaluate::AsGenericExpr(sym).value(), stmtCtx));
```

Comment:
The complexity creep of the novel handling of variables that are pointers and allocatables. Why does the symbol map for these return the _value_ of the pointer? Particularly when the code reads "value.getAddr()"?

Reply:
What do you mean by `the value of the pointer` ?

What `v.getAddr()` returned here for pointers was the address of the descriptor, which makes sense to me from a symbol table perspective, but is not what we need in this function that does not care about pointers/allocatable and just need to return the address of the variable that can be assigned to (not the address of the pointer).
`genExprAddr` seems the generic way to get this address to me (it will take care of reading the descriptor to get the address for pointer/allocatables).

Reply:
A variable of type `T` should be a location in memory allocated to be big enough to hold a `T` object. It has an address because it is a variable in memory. If `T` is a pointer, then the variable needs to be big enough to hold a value that is a pointer to `T`.

Here, we have a `getAddr()` function that returns the address of the variable of type pointer to `T`. We also have a function `genExprAddr()`, which reads almost identically, that returns the value of a pointer to T. The value of a pointer is the address of the object that is pointed at. Effectively, `genExprAddr()`, despite the name is returning a variable's value.

```c
int i;
&i ;  // i's address.
i;   // i's value.
int *p;
&p; // p's address.
p;   // p's value == the address of the pointee.
*p; // pointee of p's value.
```

In Fortran, this can be more convoluted than the C case above because the pointer is essentially a struct.

```c
struct pointer-to-int { ...; int * ref; ... } p;
&p; // p's address.
p;   // p's value != the address of the pointee.
p.ref; // the address of the pointee.
*p.ref; // pointee of p's value.
```

We seem to have `genExprAddr()` returning `&i` and `p.ref` depending on the type of the variable, while `getAddr()` returns `&i` or `&p`, which is consistent if not exactly what's needed here.

Reply:
Should we rename `SymbolBox.getAddr()` to `SymbolBox.getBase()` to avoid any confusion between the two here, because what `SymbolBox.getAddr()` is may anyway not be an address (if the symbol is directly mapped to a value, like it recently was the case in an OpenMP issue ([on flang slack channel](https://flang-compiler.slack.com/archives/C013C6AE45Q/p1636059067016900?thread_ts=1636026963.014200&cid=C013C6AE45Q))).